src/server/courseInstance/__tests__/courseInstance.service.test.ts

```diff
@@ -0,0 +1,93 @@
+import { OFFERED } from 'common/constants';
+import { CourseListingView } from 'server/course/CourseListingView.entity';
+import { Course } from 'server/course/course.entity';
+import { Faculty } from 'server/faculty/faculty.entity';
+import { CourseInstance } from '../courseinstance.entity';
+import { CourseInstanceService } from '../courseInstance.service';
+import { MultiYearPlanView } from '../MultiYearPlanView.entity';
+import { FacultyCourseInstance } from '../facultycourseinstance.entity';
+import { ScheduleBlockView } from '../ScheduleBlockView.entity';
+import { CourseInstanceListingView } from '../CourseInstanceListingView.entity';
+
+describe('Course Instance service', function () {
+  const mockRepository: Record<string, SinonStub> = {};
+  let mockInstanceRepository: Record<string, SinonStub>;
+  let courseInstanceService: CourseInstanceService;
+  beforeEach(async function () {
+    mockInstanceRepository = {
+      findOneOrFail: stub(),
+      save: stub(),
+    };
+
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        {
+          provide: getRepositoryToken(CourseListingView),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(MultiYearPlanView),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(Course),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(CourseInstance),
+          useValue: mockInstanceRepository,
+        },
+        {
+          provide: getRepositoryToken(Faculty),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(FacultyCourseInstance),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(ScheduleBlockView),
+          useValue: mockRepository,
+        },
+        {
+          provide: getRepositoryToken(CourseInstanceListingView),
+          useValue: mockRepository,
+        },
+        CourseInstanceService,
+      ],
+      controllers: [],
+    }).compile();
+
+    courseInstanceService = module
+      .get<CourseInstanceService>(CourseInstanceService);
+  });
+  describe('editCourseInstance', function () {
+    const newOfferedValue = OFFERED.N;
+    const expectedResponse = {
+      ...cs50FallInstanceUpdate,
+      offered: newOfferedValue,
+    };
+    beforeEach(function () {
+      mockInstanceRepository
+        .findOneOrFail.resolves(cs50CourseInstance);
+      mockInstanceRepository.save.resolves(expectedResponse);
+    });
+    it('returns the updated instance', async function () {
+      const result = await courseInstanceService.editCourseInstance(
+        cs50CourseInstance.id, {
+          ...cs50FallInstanceUpdate,
+          offered: newOfferedValue,
+        }
+      );
+      deepStrictEqual(result, expectedResponse);
+    });
+  });
+});
```

Comment:
As noted in the DTO, trying to update an instance's offered value to `OFFERED.BLANK` is returning a validation error, and I think that's one place where only having unit tests of the controllers/services can lead to a lot of holes in our test coverage (not the test coverage number, specifically, but just the general confidence we should have that our tests are catching errors).

I'd suggest that, instead, we probably want to use something like in `tests/server/meetings/meeting.controller.test.ts`, where we stand up the whole app and use the exposed API to send different types of test data (i.e. all of the different `OFFERED` values), then confirm that 1) the correct data is returned, and 2) the correct data is persisted to the database.

Reply:
Thanks for catching this! I added tests for all the different options through [60dc4a3](https://github.com/seas-computing/course-planner/pull/473/commits/60dc4a325b8119b5f3d115cc264c94b986f9e166) so that we we checking each of the possible values for `OFFERED`.